pkg/registration/hub/lease/clocksynccontroller.go

```diff
@@ -0,0 +1,200 @@
+	clusterv1informer "open-cluster-management.io/api/client/cluster/informers/externalversions/cluster/v1"
+	clusterv1listers "open-cluster-management.io/api/client/cluster/listers/cluster/v1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+
+	"open-cluster-management.io/ocm/pkg/common/patcher"
+	"open-cluster-management.io/ocm/pkg/common/queue"
+)
+
+type clusterLease struct {
+	savedLease *coordv1.Lease
+}
+
+func (h *clusterLease) deepCopy() *clusterLease {
+	if h == nil {
+		return nil
+	}
+	return &clusterLease{
+		savedLease: h.savedLease.DeepCopy(),
+	}
+}
+
+type clusterLeaseMap struct {
+	sync.RWMutex
+	clusterLeases map[string]*clusterLease
+}
+
+func (m *clusterLeaseMap) getDeepCopy(clusterName string) *clusterLease {
+	m.RLock()
+	defer m.RUnlock()
+	return m.clusterLeases[clusterName].deepCopy()
+}
+
+func (m *clusterLeaseMap) set(clusterName string, clusterHealth *clusterLease) {
+	m.Lock()
+	defer m.Unlock()
+	m.clusterLeases[clusterName] = clusterHealth
+}
+
+type clockSyncController struct {
+	patcher       patcher.Patcher[*clusterv1.ManagedCluster, clusterv1.ManagedClusterSpec, clusterv1.ManagedClusterStatus]
+	clusterLister clusterv1listers.ManagedClusterLister
+	leaseLister   coordlisters.LeaseLister
+	eventRecorder events.Recorder
+
+	clusterLeaseMap *clusterLeaseMap
+}
+
+const (
+	clockSyncControllerName = "ClockSyncController"
+)
+
+func NewClockSyncController(
+	clusterClient clientset.Interface,
+	clusterInformer clusterv1informer.ManagedClusterInformer,
+	leaseInformer coordinformers.LeaseInformer,
+	recorder events.Recorder,
+) factory.Controller {
+	c := &clockSyncController{
+		patcher: patcher.NewPatcher[
+			*clusterv1.ManagedCluster, clusterv1.ManagedClusterSpec, clusterv1.ManagedClusterStatus](
+			clusterClient.ClusterV1().ManagedClusters()),
+		clusterLister: clusterInformer.Lister(),
+		leaseLister:   leaseInformer.Lister(),
+		eventRecorder: recorder.WithComponentSuffix("managed-cluster-clock-sync-controller"),
+		clusterLeaseMap: &clusterLeaseMap{
+			clusterLeases: make(map[string]*clusterLease),
+		},
+	}
+
+	syncCtx := factory.NewSyncContext(clockSyncControllerName, recorder)
+	leaseRenewTimeUpdateInformer := renewUpdateInfomer(syncCtx.Queue(), leaseInformer)
+
+	return factory.New().WithSyncContext(syncCtx).
+		WithFilteredEventsInformersQueueKeysFunc(
+			queue.QueueKeyByLabel(clusterv1.ClusterNameLabelKey),
+			queue.UnionFilter(queue.FileterByLabel(clusterv1.ClusterNameLabelKey), queue.FilterByNames(leaseName)),
+			leaseRenewTimeUpdateInformer,
+		).
+		WithSync(c.sync).
+		ToController(clockSyncControllerName, recorder)
+}
+
+func renewUpdateInfomer(queue workqueue.RateLimitingInterface, leaseInformer coordinformers.LeaseInformer) cache.SharedIndexInformer {
+	leaseRenewTimeUpdateInformer := leaseInformer.Informer()
+	_, err := leaseRenewTimeUpdateInformer.AddEventHandler(&cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(oldObj, newObj interface{}) {
```

Comment:
Only the renew update event will trigger the reconcile, but I'm afriad of the default 10 hours reconcile of controller-runtime will cause a incorrect reconcile. That's why we still need the local cache. 
@qiujian16

Reply:
I am not sure I follow. I do not think it needs to care the resync, the reconile should be only triggered upon update event.

Reply:
Got it, the controller updated.